### PR TITLE
[core] Consolidate compiled-kernel caches under SGLANG_CACHE_DIR

### PR DESCRIPTION
**Pull request by shuwang21**
## Motivation (Idea from @ch-wan)

Triton, Inductor, FlashInfer and the CUDA driver each write their compiled
kernels to their own directory:

```
~/.triton/                    
~/.cache/flashinfer/
/tmp/torchinductor_$USER/     
~/.nv/
```
Scattered across four places, they cannot be cleaned, warmed, copied between
hosts or volume-mounted as a unit, and nothing tells an operator where they
all are.

## Modifications

- **`environ.py`** — added `third_party_cache_defaults()` (env var → cache dir
  mapping) and `redirect_third_party_caches()`, which `setdefault`s those four
  vars. `SGLANG_DG_CACHE_DIR` now defaults to `{SGLANG_CACHE_DIR}/deep_gemm`.
- **`sglang/__init__.py`** — calls the redirect above the imports that reach
  torch and FlashInfer, which otherwise fix these paths first.
- **`compile_utils.py`** — reads `SGLANG_DG_CACHE_DIR` through `envs` instead
  of duplicating the default.
- **CI scripts** — `warmup_deep_gemm.py`'s `MARKER_DIR` derives from
  `SGLANG_CACHE_DIR`; `ci_install_dependency.sh` clears the inductor cache in
  every location it can be in.
- **Diffusion** — `_configure_persistent_torch_compile_cache` compares against
  `third_party_cache_defaults()`, so an sglang default is not mistaken for an
  operator override and its caches still land under
  `SGLANG_DIFFUSION_CACHE_ROOT`.
- **Startup / docs** — logs `~/.triton`, `~/.cache/flashinfer` and
  `~/.cache/deep_gemm` if present, without touching them; environment-variable
  doc updated.

> [!IMPORTANT]
> Every cache moves, so the first launch after upgrading recompiles once.
> Copy or symlink the old directory to the new one if you pre-warm or
> volume-mount it — `deep_gemm` and `flashinfer` are the expensive ones.
>
> | was | now |
> |---|---|
> | `~/.triton` | `{SGLANG_CACHE_DIR}/triton` |
> | `~/.cache/flashinfer` | `{SGLANG_CACHE_DIR}/.cache/flashinfer` |
> | `~/.cache/deep_gemm` | `{SGLANG_CACHE_DIR}/deep_gemm` |
> | `/tmp/torchinductor_$USER` | `{SGLANG_CACHE_DIR}/inductor` |
> | `~/.nv/ComputeCache` | `{SGLANG_CACHE_DIR}/nv` |
>
> Setting any of `TRITON_CACHE_DIR`, `FLASHINFER_WORKSPACE_BASE`,
> `SGLANG_DG_CACHE_DIR`, `TORCHINDUCTOR_CACHE_DIR` or `CUDA_CACHE_PATH`
> keeps that one where it is.

## Verification

E2E on a clean L4 studio, Qwen2.5-1.5B:

- all four env vars resolve under `~/.cache/sglang`
- after a full launch + request, every cache landed there
  (`triton` 464K, `.cache/flashinfer` 5.9M, `inductor`, `nv`) and nothing
  appeared outside it
- the legacy-dir notice fires only when those directories exist

## Accuracy Tests

No forward or weight-loading logic changed — only where derived artifacts are
written.

## Speed Tests and Profiling

No inference-path impact. One-time recompile on upgrade, see the note above.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #30972987487](https://github.com/sgl-project/sglang/actions/runs/30972987487)
Latest PR Test (Extra): :white_check_mark: [Run #30972987297](https://github.com/sgl-project/sglang/actions/runs/30972987297)

### CODE REVIEW

**Comment by shuwang21** [comment]
one open question: should presharded checkpoints really live under the same root as the kernel caches?

File: python/sglang/srt/model_loader/loader.py [hunk 1]
```diff
@@ -1720,7 +1721,16 @@ def _presharded_dir(
...
+            # <SGLANG_CACHE_DIR>/presharded/<model_id>: the cache dir is shared
+            # across models, so the root needs a per-model level of its own.
+            model_id = hashlib.sha256(
+                str(model_config.model_path).encode()
+            ).hexdigest()[:16]
+            root = os.path.join(
+                os.path.expanduser(envs.SGLANG_CACHE_DIR.get()),
+                self.DEFAULT_SUBDIR,
+                model_id,
+            )
```
[the remark this answers is not in the document]
Comment (shuwang21) [reply python/sglang/srt/model_loader/loader.py]: nice catch, will fix this in [the follow up pr ](https://github.com/sgl-project/sglang/pull/32437)

File: python/sglang/srt/model_loader/loader.py [hunk 2]
```diff
@@ -1720,7 +1721,16 @@ def _presharded_dir(
         else:
             root = self._presharded_path_override
         if root is None:
-            root = os.path.join(model_config.model_path, self.DEFAULT_SUBDIR)
+            # <SGLANG_CACHE_DIR>/presharded/<model_id>: the cache dir is shared
+            # across models, so the root needs a per-model level of its own.
+            model_id = hashlib.sha256(
+                str(model_config.model_path).encode()
+            ).hexdigest()[:16]
```
[the remark this answers is not in the document]
Comment (shuwang21) [reply python/sglang/srt/model_loader/loader.py]: nice catch, will fix this in [the follow up pr ](https://github.com/sgl-project/sglang/pull/32437)

File: python/sglang/srt/environ.py [hunk 3]
```diff
@@ -1212,6 +1222,28 @@ def _warn_deprecated_env_to_cli_flag(env_name: str, suggestion: str):
...
+
+    Must run before torch and FlashInfer are imported: FlashInfer resolves its
+    workspace at import time, and torch writes TORCHINDUCTOR_CACHE_DIR into the
+    environment itself, so a later setdefault is a no-op. sglang/__init__.py
+    imports this module first for that reason.
+    """
+    base = os.path.expanduser(envs.SGLANG_CACHE_DIR.get())
+    defaults = {
+        "TRITON_CACHE_DIR": os.path.join(base, "triton"),
+        "TORCHINDUCTOR_CACHE_DIR": os.path.join(base, "inductor"),
```
Comment (ch-wan) [inline python/sglang/srt/environ.py:1383]: **[bug]** Setting `TORCHINDUCTOR_CACHE_DIR` / `TRITON_CACHE_DIR` via `os.environ.setdefault` at `import sglang` breaks multimodal diffusion’s intended cache isolation. `gpu_worker._configure_persistent_torch_compile_cache` only rewrites those env vars when they are unset or under `tempfile.gettempdir()`; any non-tmp value is treated as an explicit user override and left alone. After this PR every diffusion entry that loads `sglang` (including `sglang serve` for diffusion models) already has non-tmp defaults under `SGLANG_CACHE_DIR`, so the worker never redirects to `SGLANG_DIFFUSION_CACHE_ROOT/torch_compile_cache`. Diffusion compile caches silently move to `~/.cache/sglang/{inductor,triton}` and wiping `SGLANG_DIFFUSION_CACHE_ROOT` no longer clears them.

**Suggestion:** Either (a) teach `_configure_persistent_torch_compile_cache` to treat the `SGLANG_CACHE_DIR`-derived paths as sglang defaults (not user overrides) and still rehome under `SGLANG_DIFFUSION_CACHE_ROOT`, (b) skip setting Inductor/Triton in `_redirect_third_party_caches` when the process is a diffusion entrypoint, or (c) deliberately unify diffusion onto `SGLANG_CACHE_DIR` and update the diffusion cache helper + docs accordingly. (a) or (c) with an explicit decision is preferred.

File: python/sglang/srt/environ.py [hunk 4]
```diff
@@ -692,7 +701,8 @@ class Envs:
     SGLANG_JIT_DEEPGEMM_FAST_WARMUP = EnvBool(False)
     SGLANG_JIT_DEEPGEMM_COMPILE_WORKERS = EnvInt(4)
     SGLANG_IN_DEEPGEMM_PRECOMPILE_STAGE = EnvBool(False)
```
Comment (ch-wan) [inline python/sglang/srt/environ.py:746]: **[suggestion]** Changing the default of `SGLANG_DG_CACHE_DIR` from `~/.cache/deep_gemm` to `{SGLANG_CACHE_DIR}/deep_gemm` orphans existing warm DeepGEMM caches and forces a one-time recompile (same class of cost the PR already calls out for Triton/FlashInfer). The in-repo docs still advertise the old default (`docs_new/docs/references/environment_variables.mdx` lists `~/.cache/deep_gemm`), so operators who pre-warm or volume-mount the old path will get a silent cold start with no doc signal.

**Suggestion:** Update the environment-variables doc (and any ops/runbook that hard-codes `~/.cache/deep_gemm`) to the new default, and mention the one-time recompile / optional copy from the legacy path in the PR or release notes.

File: scripts/ci/cuda/warmup_deep_gemm.py [hunk 5]
```diff
@@ -26,8 +26,8 @@
 from pathlib import Path
 from typing import Dict, List
 
-# Shared with warmup_server.py. Wipe alongside /root/.cache/deep_gemm if you
-# clear the DeepGEMM JIT cache — a stale marker → in-test JIT compile.
+# Shared with warmup_server.py. Wipe alongside the DeepGEMM JIT cache under
+# SGLANG_CACHE_DIR if you clear it — a stale marker → in-test JIT compile.
 MARKER_DIR = os.path.join(os.path.expanduser("~"), ".cache", "sglang", "warmup_markers")
```
Comment (ch-wan) [inline scripts/ci/cuda/warmup_deep_gemm.py]: **[suggestion]** The comment now says to wipe markers “alongside the DeepGEMM JIT cache under `SGLANG_CACHE_DIR`”, but `MARKER_DIR` is still hard-coded to `~/.cache/sglang/warmup_markers`. If CI or a developer overrides `SGLANG_CACHE_DIR`, the warmed `DG_JIT_CACHE_DIR` and the marker tree diverge: markers may claim “warmed” while the server’s actual cache dir is empty (or the reverse), reintroducing the stale-marker / cold-compile failure mode this script exists to prevent.

**Suggestion:** Derive `MARKER_DIR` from `envs.SGLANG_CACHE_DIR.get()` (e.g. `os.path.join(..., "warmup_markers")`) so it tracks the same root as `SGLANG_DG_CACHE_DIR`.

File: python/sglang/__init__.py [hunk 6]
```diff
@@ -1,5 +1,13 @@
 # SGLang public APIs
 
+# Imported for its side effect of routing third-party JIT caches under
+# SGLANG_CACHE_DIR. Must come before anything that pulls in torch, which writes
+# TORCHINDUCTOR_CACHE_DIR into the environment itself at import time and would
```
Comment (ch-wan) [inline python/sglang/__init__.py]: **[nit]** The comment claims torch “writes `TORCHINDUCTOR_CACHE_DIR` into the environment itself at import time”. In current PyTorch (`torch/_inductor/runtime/cache_dir_utils.py`), that write happens on the first `cache_dir()` call, not at `import torch`. FlashInfer’s workspace resolution at import *is* import-time sensitive; Inductor is “before first inductor use”. The early import of `environ` is still the right fix, but the stated mechanism is slightly wrong and could mislead a future “optimization” that moves the redirect later.

**Suggestion:** Rephrase to something like: “Must run before FlashInfer is imported (workspace is resolved at import) and before the first Inductor `cache_dir()` call (which setdefaults `TORCHINDUCTOR_CACHE_DIR` and would make a later setdefault a no-op).”

File: python/sglang/__init__.py [hunk 7]
```diff
@@ -1,5 +1,13 @@
...
+# otherwise win. The module is stdlib-only, so importing it here is cheap.
```
Comment (ch-wan) [inline python/sglang/__init__.py]: **[nit]** The comment says `environ` is “stdlib-only”, but the module still import-side-effects `sglang.srt.debug_utils.cuda_coredump` at the bottom (and pulls `subprocess`). That path is still cheap and torch-free, so the placement is fine; the claim is just overstated.

**Suggestion:** Soften to “does not import torch / third-party heavy deps” (or drop the purity claim).

File: test/registered/unit/test_environ_cache_dirs.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
+"""Tests for routing third-party JIT caches under SGLANG_CACHE_DIR.
```
Comment (Fridge003) [inline test/registered/unit/test_environ_cache_dirs.py]: No need to add this test

File: python/sglang/__init__.py [hunk 9]
```diff
@@ -1,5 +1,13 @@
...
+from sglang.srt import environ as _environ
+
+del _environ
```
Comment (Fridge003) [inline python/sglang/__init__.py]: Deleting environ is a little confusing. Can we call `redirect_third_party_caches` here instead.

File: test/registered/unit/test_environ_cache_dirs.py (hunk 8 above)
Comment (shuwang21) [reply test/registered/unit/test_environ_cache_dirs.py]: Removed.

File: python/sglang/__init__.py (hunk 9 above)
Comment (shuwang21) [reply python/sglang/__init__.py]: Sure. Replaced with an explicit `redirect_third_party_caches()` call, and dropped the module-level side effect in `environ.py` so it only runs from that one call site.

File: python/sglang/__init__.py (hunk 7 above)
Comment (shuwang21) [reply python/sglang/__init__.py]: Softened to "no heavy dependency (no torch)".

File: python/sglang/__init__.py (hunk 6 above)
Comment (shuwang21) [reply python/sglang/__init__.py]: Corrected and rephrased inside `redirect_third_party_caches` about this requirement.

File: python/sglang/srt/environ.py (hunk 3 above)
Comment (shuwang21) [reply python/sglang/srt/environ.py:1383]: Nice catch. Went with (a).

Extracted the mapping into `third_party_cache_defaults()`. The diffusion helper now compares against it exactly.

Diffusion keeps its caches under `SGLANG_DIFFUSION_CACHE_ROOT`.

File: scripts/ci/cuda/warmup_deep_gemm.py (hunk 5 above)
Comment (shuwang21) [reply scripts/ci/cuda/warmup_deep_gemm.py]: Nice catch. Fixed. `MARKER_DIR` now derives from `envs.SGLANG_CACHE_DIR.get()`.

Chasing this found one more of the same class: `ci_install_dependency.sh` clears the inductor cache and was still targeting `/tmp/torchinductor_$USER`, so that cleanup had silently become a no-op. It now clears every location the cache can be in.

File: python/sglang/srt/environ.py (hunk 4 above)
Comment (shuwang21) [reply python/sglang/srt/environ.py:746]: Updated the docs. `SGLANG_DG_CACHE_DIR` now lists `{SGLANG_CACHE_DIR}/deep_gemm`, and the `SGLANG_CACHE_DIR` entry says it is the root for compiled-kernel caches.

Also added `~/.cache/deep_gemm` to the startup notice about orphaned cache dirs, and called out the one-time recompile in the PR body.

**Review by Fridge003 — APPROVED** [review]

**Comment by shuwang21** [comment]
Rebased to resolve conflicts.
